src/ballet/bmtree/fd_bmtree.h

```diff
@@ -0,0 +1,284 @@
+
+     4. Construct full tree:
+
+        (vector of leaf blobs) -> (tree of nodes)
+
+     5. Create inclusion proof from tree data
+
+        (tree of nodes, node index) -> (inclusion proof)
+
+     6. Verify node inclusion proof
+
+        (node, root node, inclusion proof) -> (bool)
+
+   **Topology**
+
+   Tree topology has the following constraints:
+
+    - All leaf nodes are in the bottom level
+
+    - If a given layer `L`
+      with number of nodes `L_n` ...
+
+      ... has exactly one node,
+          this one node contains the root node
+          and is the uppermost layer.
+
+      ... has more than one node
+          ... and `L_n % 2 == 0`,
+              the layer above contains n/2 nodes
+
+          ... and `L_n % 2 == 1`,
+              the layer above contains (n/2)+1 nodes.
+
+   Example: Tree with 1 leaf node (root node = leaf node)
+
+    L0
+
+   Example: Tree with 4 leaf nodes
+
+           Iδ
+          /  \
+         /    \
+       Iα      Iβ
+      /  \    /  \
+     L0  L1  L2  L3
+
+   Example: Tree with 5 leaf nodes
+
+              Iζ
+             /  \
+            /    \
+           Iδ     Iε
+          /  \     \\
+         /    \     \\
+       Iα      Iβ    Iγ
+      /  \    /  \   ||
+     L0  L1  L2  L3  L4
+
+   **Construction**
+
+   First, each input in the vector (an arbitrary-sized binary blob)
+   is hashed to leaf node.
+
+   Then, the hash function is recursively applied over pairs of nodes
+   until there is only one node left (the root).
+
+   `fd_bmtree_32` uses the full SHA-256 digest for each tree node
+   and is thus considered cryptographically secure.
+
+   `fd_bmtree_20` uses SHA-256 digests truncated to 160 bits.
+
+   **Inclusion Proofs**
+
+   Inclusion proofs are used to verify whether a set of leaf nodes is
+   part of a commitment (identified by the root node).
+
+   At a high level, inclusion proofs present a sequence of hash
+   instructions that when executed result in the root commitment.
+
+   Inclusion proof size is O(log n) with regards to tree node count.
+
+   Various types of inclusion proofs exist:
+
+     - Single inclusion proofs (over one leaf node)
+     - Range inclusion proofs (over a contiguous range of leaf nodes)
+     - Sparse inclusion proofs (over an arbitrary subset of leaf nodes) */
+
+#define FD_BMTREE_PREFIX_LEAF   (uchar)0x00
+#define FD_BMTREE_PREFIX_BRANCH (uchar)0x01
+
+/* FD_BMTREE20_NODE_SZ: Size in bytes of a SHA-256-160 tree node */
+#define FD_BMTREE20_NODE_SZ (20UL)
+
+/* FD_BMTREE32_NODE_SZ: Size in bytes of a SHA-256 tree node */
+#define FD_BMTREE32_NODE_SZ (32UL)
+
+/* fd_bmtree20_node_t is the hash of a SHA-256-160 tree node (20 bytes) */
+typedef uchar fd_bmtree20_node_t[FD_BMTREE20_NODE_SZ];
+
+/* fd_bmtree20_node_t is the hash of a SHA-256 tree node (32 bytes) */
+typedef uchar fd_bmtree32_node_t[FD_BMTREE32_NODE_SZ];
+
+/* fd_bmtree*_commit_t computes the root of a binary Merkle tree.
```

Comment:
I'm confused by a data structure being described by a verb. How about this?

`/* fd_bmtree*_commit_t stores intermediate state used to compute the root of a binary Merkle tree built incrementally.`